lib/ansible/plugins/filter/core.py

```diff
@@ -294,25 +293,93 @@ def mandatory(a):
 
 
 def combine(*terms, **kwargs):
-    recursive = kwargs.get('recursive', False)
-    if len(kwargs) > 1 or (len(kwargs) == 1 and 'recursive' not in kwargs):
-        raise AnsibleFilterError("'recursive' is the only valid keyword argument")
-
-    dicts = []
-    for t in terms:
-        if isinstance(t, MutableMapping):
-            recursive_check_defined(t)
-            dicts.append(t)
-        elif isinstance(t, list):
-            recursive_check_defined(t)
-            dicts.append(combine(*t, **kwargs))
-        else:
-            raise AnsibleFilterError("|combine expects dictionaries, got " + repr(t))
-
-    if recursive:
-        return reduce(merge_hash, dicts)
-    else:
-        return dict(itertools.chain(*map(iteritems, dicts)))
+    recursive = kwargs.pop('recursive', False)
+    list_merge = kwargs.pop('list_merge', 'keep')
+    if kwargs:
+        raise AnsibleFilterError("'recursive' and 'list_merge' are the only valid keyword argument")
+    if list_merge not in ('keep', 'append', 'prepend', 'append_np', 'prepend_np'):
+        raise AnsibleFilterError("'list_merge' can only be equal to 'keep', 'append', 'prepend', 'append_np' or 'prepend_np'")
+
+    # allow the user to do `[dict1, dict2, ...] | combine`
+    dicts = flatten(terms, levels=1)
+
+    # recursively check that every elements are defined (for jinja2)
+    recursive_check_defined(dicts)
+
+    # function that will be use to merge 2 dicts
+    def _combine_two_dicts(x, y):
```

Comment:
you seem to be re-implementing merge_hash, just update the function instead

Reply:
done